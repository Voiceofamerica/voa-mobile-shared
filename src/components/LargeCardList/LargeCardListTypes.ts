
import * as React from 'react'

import { ListItem } from '../../helpers/itemList'

export { ListItem }

export interface IconDefinition {
  icon: string | JSX.Element
  getClass?: (id: number) => string
  getStyle?: (id: number) => React.CSSProperties
  onPress: (id: number) => void
}

export interface BaseProps {
  items: ListItem[] | undefined
  onItemClick: (id: number) => void
  className?: string
  style?: React.CSSProperties
  iconButtons?: IconDefinition[]
}
