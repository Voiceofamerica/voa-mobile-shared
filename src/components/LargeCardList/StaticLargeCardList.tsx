
import * as React from 'react'

import LargeCard, { IconDefinition } from '../LargeCard'

import { themed, ThemeProps, DEFAULT_THEME } from '../ThemeProvider'
import { BaseProps } from './LargeCardListTypes'
import { cardList, row, fixed } from './LargeCardList.scss'

export interface Props extends BaseProps, ThemeProps {
}

@themed
class StaticCardList extends React.Component<Props> {
  render () {
    const { items = [], theme = DEFAULT_THEME } = this.props
    const {
      largeCardListBackground,
    } = theme

    return (
      <div className={cardList} style={{ background: largeCardListBackground }}>
        {
          items.length > 0
          ? this.renderContent()
          : ''
        }
      </div>
    )
  }

  private renderContent = () => {
    const { items = [], onItemClick } = this.props

    return items.map(({ id, image, title, icon }) => (
      <div key={id} className={`${fixed} ${row}`}>
        <LargeCard
          onPress={() => onItemClick(id)}
          title={title}
          imageUrl={image && image.hero}
          titleIcon={icon}
          iconButtons={this.makeIconButtons(id)}
        />
      </div>
    ))
  }

  private makeIconButtons = (id: number): IconDefinition[] => {
    const { iconButtons = [] } = this.props
    return iconButtons.map(({ icon, onPress, getClass, getStyle }) => ({
      icon,
      getClassName: getClass ? () => getClass(id) : undefined,
      getStyle: getStyle ? () => getStyle(id) : undefined,
      onPress: () => onPress(id),
    }))
  }
}

export default StaticCardList
