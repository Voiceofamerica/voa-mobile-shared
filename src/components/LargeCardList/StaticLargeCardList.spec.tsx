
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import { ListItem, IconDefinition } from './LargeCardListTypes'
import StaticLargeCardList from './StaticLargeCardList'

jest.mock('../LargeCard')
jest.mock('../SvgIcon')

const testItem: ListItem = {
  id: 1,
  image: {
    hero: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w500_r1.png',
  },
  minorText: '15:08',
  title: 'President Trump: From North Korea to Iran, we confront dangerous governments',
}

const createList = (count: number) => {
  const output: ListItem[] = []
  for (let i = 0; i < count; i++) {
    output.push({
      ...testItem,
      id: i,
    })
  }
  return output
}

const list = createList(10)

describe(`<${StaticLargeCardList.name} />`, () => {
  describe('shapshots', () => {
    it('should render all of the items', () => {
      const element = create((
        <StaticLargeCardList items={list} onItemClick={jest.fn()} />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should render with icon buttons', () => {
      const icons: IconDefinition[] = [
        {
          icon: 'icon 1',
          onPress: jest.fn(),
        },
        {
          icon: 'icon 2',
          onPress: jest.fn(),
        },
        {
          icon: 'icon 3',
          onPress: jest.fn(),
        },
      ]
      const element = create((
        <StaticLargeCardList items={list} onItemClick={jest.fn()} iconButtons={icons} />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should render nothing if no items', () => {
      const element = create((
        <StaticLargeCardList items={[]} onItemClick={jest.fn()} />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should not throw if items is undefined', () => {
      const element = create((
        <StaticLargeCardList items={undefined} onItemClick={jest.fn()} />
      ))

      expect(element).toMatchSnapshot()
    })
  })
})
