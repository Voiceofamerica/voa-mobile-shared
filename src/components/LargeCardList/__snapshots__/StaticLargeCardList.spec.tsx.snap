// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StaticCardList /> shapshots should not throw if items is undefined 1`] = `
<div
  className="cardList"
>
  
</div>
`;

exports[`<StaticCardList /> shapshots should render all of the items 1`] = `
<div
  className="cardList"
>
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
</div>
`;

exports[`<StaticCardList /> shapshots should render nothing if no items 1`] = `
<div
  className="cardList"
>
  
</div>
`;

exports[`<StaticCardList /> shapshots should render with icon buttons 1`] = `
<div
  className="cardList"
>
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
  <div
    className="fixed row"
  />
</div>
`;
